File: q6-valid-palindrome/q6-sol.js
const subIsPalindrome = function(s, p1, p2) {
    while (p1 < p2) {
        if (s[p1] !== s[p2]) return false;
        p1++;
        p2--
    }
    return true;
}

const validPalindrome = function(s) {
    if (s.length <= 1) return true;
    
    //s = s.replace(/[^A-Za-z0-9]/g, '').toLowerCase();
    let p1 = 0, p2 = s.length - 1;
    
    while (p1 <= p2 - 1) {
        if (s[p1] !== s[p2]) {
            return (subIsPalindrome(s, p1+1, p2) || subIsPalindrome(s, p1, p2-1));
        }
        p1++;
        p2--;
    }
    
    return true;
};

// time: O(n)
// space: O(1)

File: q6-almost-valid-palindrome/q6-sol.cpp
// LeetCode 680.
// https://leetcode.com/problems/valid-palindrome-ii/

// Test Cases:
// s = "aba" -> true
// s = "raceacar" -> true
// s = "a" -> true
// s = "ab" -> true

// "abxxxfba"
//    i  j
// remove s[i] -> "abxxfba"  if 2nd round s[i] != s[j] -> check the other
//                   i j 
// remove s[j] -> "abxxxba"  --> if still not eaqul -> return false
//                   i j 
#include <iostream>
#include <string>

using namespace std;

class Solution {
public:
    bool checkPalindrome(string s, int i, int j) {
        while (i < j) {
            if (s[i] != s[j]) return false;
            i++;
            j--;
        }
        return true;
    }
    
    bool validPalindrome(string s) {
        if(s.length() <= 2) return true;
        
        int i = 0, j = s.length() - 1;
        
        string temp = s;
        
        while (i < j) {
            
            if (s[i] != s[j]) {
                return checkPalindrome(s, i + 1, j) || checkPalindrome(s, i, j - 1);
            }
            i++;
            j--;
        }
        
        return true;
    }
};

// time: O(n)
// space: O(1)